data_model/src/lib.rs

```diff
@@ -1005,19 +1006,214 @@ impl From<LengthLimits> for RangeInclusive<usize> {
+        }
+    }
+
+    #[doc(hidden)]
+    #[must_use]
+    fn unwrapped_or(self, other: Self) -> Self {
+        match self.or(other) {
+            ControlFlow::Continue(val) | ControlFlow::Break(val) => val,
+        }
+    }
+
+    /// Given a list of all possible values of a type implementing [`PredicateSymbol`]
+    /// which are different in predicate context, exhaustively tests for:
+    /// - commutativity of `and` and `or`
+    /// - associativity of `and` and `or`
+    /// - De Mornan duality of `and` and `or`
+    /// - double negation elimination
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use iroha_data_model::PredicateSymbol;
+    ///
+    /// fn test() {
+    ///     PredicateSymbol::test_conformity(vec![true, false]);
+    /// }
+    /// ```
+    ///
+    fn test_conformity(values: Vec<Self>)
+    where
+        Self: PartialEq + Clone,
+    {
+        Self::test_conformity_with_eq(values, <Self as PartialEq>::eq);
+    }
+
+    /// Same as [`test_conformity`](`PredicateSymbol::test_conformity`), but
+    /// if type implementing [`PredicateSymbol`] carries some internal state
+    /// that isn't associative, one can provide custom `shallow_eq` function
+    /// that will be called instead of [`PartialEq::eq`]
+    ///
+    /// # Examples
+    ///
+    ///
+    /// ```rust
+    /// use std::ops::ControlFlow;
+    /// use iroha_data_model::PredicateSymbol;
+    ///
+    /// #[derive(Clone, PartialEq)]
+    /// enum Check {
+    ///    Good,
+    ///    // Encapsulates reason for badness which
+    ///    // doesn't behave associatively
+    ///    // (but if we ignore it, Check as a whole does)
+    ///    Bad(String),
+    /// }
+    ///
+    /// impl core::ops::Not for Check {
+    ///   type Output = Self;
+    ///   fn not(self) -> Self {
+    ///     // ...
+    ///     todo!()
+    ///   }
+    /// }
+    ///
+    /// impl PredicateSymbol for Check {
+    ///   fn and(self, other: Self) -> ControlFlow<Self, Self> {
+    ///     // ...
+    ///     todo!()
+    ///   }
+    ///
+    ///   fn or(self, other: Self) -> ControlFlow<Self, Self> {
+    ///     // ...
+    ///     todo!()
+    ///   }
+    /// }
+    ///
+    /// fn shallow_eq(left: &Check, right: &Check) -> bool {
+    ///    match (left, right) {
+    ///      (Check::Good, Check::Good) | (Check::Bad(_), Check::Bad(_)) => true,
+    ///      _ => false
+    ///    }
+    /// }
+    ///
+    /// #[test]
+    /// fn test() {
+    ///    let good = Check::Good;
+    ///    let bad = Check::Bad("example".to_owned());
+    ///    // Would fail some assertions, since derived PartialEq is "deep"
+    ///    // PredicateSymbol::test_conformity(vec![good, bad]);
+    ///
+    ///    // Works as expected
+    ///    PredicateSymbol::test_conformity_with_eq(vec![good, bad], shallow_eq);
+    /// }
+    /// ```
+    fn test_conformity_with_eq(values: Vec<Self>, shallow_eq: impl FnMut(&Self, &Self) -> bool)
```

Comment:
`shallow_eq` could be `impl Fn(&Self, &Self) -> bool` or even `fn(&Self, &Self) -> bool` to me it's seems unlikely that we ever want to capture context with `shallow_eq` function. 

But i will approve regardless.

Reply:
Usually I just choose the bare minimum I need/will plausibly need not to add any additional restrictions on the caller.
Are there any benefits in further restricting `shallow_eq`?

Reply:
> Usually I just choose the bare minimum I need/will plausibly need not to add any additional restrictions on the caller.

Sounds like reassuring strategy. 

> Are there any benefits in further restricting `shallow_eq`?

Compile time if we are talking about fn pointers.